classpath/java/util/ArrayList.java

```diff
@@ -183,6 +183,20 @@ public void clear() {
   public ListIterator<T> listIterator() {
     return listIterator(0);
   }
+  
+  @Override
+  public boolean retainAll(Collection<?> c) {
+    boolean changed = false;
+    Iterator<T> iter = this.iterator();
+    while(iter.hasNext()) {
+      T item = iter.next();
+      if(!c.contains(item)) {
+        changed = true;
+        iter.remove();
```

Comment:
The idea behind retainAll (i believe) is to only retain items in the collection backing collection that are in the passed in collection, so any/all items in the backing collection that are not in the passed in one need to be removed.  We do not technically need to set changed to true after each one, but its not worth a conditional each time.  We do however have to go through the whole list of items in the backing collection.